bookkeeper-server/src/test/java/org/apache/bookkeeper/client/BookieDecommissionTest.java

```diff
@@ -44,14 +46,17 @@
 
     public BookieDecommissionTest() {
         super(NUM_OF_BOOKIES, 480);
-        baseConf.setOpenLedgerRereplicationGracePeriod(String.valueOf(30000));
-        setAutoRecoveryEnabled(true);
+        baseConf.setOpenLedgerRereplicationGracePeriod(100);
+        setAutoRecoveryEnabled(false);
```

Comment:
I'm not sure how can this be ok? you disabled AutoRecoveryService for all the bookies by setting this to false, but in each testcase you are starting  AutoRecoveryMain (which starts daemon for Auditor and ReplicationWorker) just for one bookie. How come having just one bookie ReplicationWorker would be sufficient for these testcases where we are killing bookie/s? 

My concern here is if the bookie which is having AutoRecoveryMain is already having copy of fragments of that ledger, then it would ignore replicating that under replicated ledger to that Bookie, right?

Reply:
the way I see this - what decommission test is doing here is to testing the decommission behavior. we are not testing killing auto recovery. we don't need to start autorecovery on every bookie, it is resource consumed and don't have more meanings to the test itself. move auto recovery process out of the bookies, it can make the auto recovery trigger more deterministic, rather than replying on non-deterministic sleeping behavior.

Reply:
I dont get your point  - "we don't need to start autorecovery on every bookie"

lets say there are 6 Bookies - BK0, BK1, BK2,..BK5 and Ledger with quorum 3,3 is created.
you started AutoRecovery process for only one Bookie - BK0. So for BK0 Auditor and ReplicationWorker daemons are running.

LID-P (ledger) : BK0, BK1, BK2 (ensemble)

Now if BK1 is killed, so LID-P will be underreplicated. My question is though AutoRecovery (Auditor, ReplicationWorker) are running for BK0, LID-P will remain under-replicated, right? Since, BK0 is already part of ensemble of LID-P, ReplicationWorker running with BK0 wouldn't be able to do anything.

Reply:
@reddycharan I think you have a wrong understanding on how autorecovery works. autorecovery is a completely isolated thing from bookies, it doesn't really stick to any bookies. It can be co-run with bookies, or it can be run as separate service. that means you can run less or more autorecovery instances that bookies.

in your example, if BK1 is killed, autorecovery daemon will detect that and choose a bookie to replace BK1 and replicate entries from BK0 and BK2 to the newly chosen bookie.

Reply:
ok..had private chat with Sijie. https://github.com/apache/bookkeeper/commit/6fcabfc80c5bd6cdfa9252cdfc4ed209cc0620a6 change has decoupled ReplicationWorker and Bookie completely, and now there is no "targetBookie" factor in ReplicationWorker. So I'm good with this change. On a different note, ReplicationWorker.java javadoc is not updated with that change and it is still using "targetBookie" in its comments.